File: cups.js
var VOLUME_X = 4;
var VOLUME_Y = 9;
var TARGET_VOLUME = 2;

if (process.argv.length >= 5) {
    VOLUME_X = parseInt(process.argv[2]);
    VOLUME_Y = parseInt(process.argv[3]);
    TARGET_VOLUME = parseInt(process.argv[4]);
}

function Cups(x, y) {
    this.x = x;
    this.y = y;

    if (typeof this.constructor.prototype.fill_x != 'function') {
        this.constructor.prototype.toString = function() {
            return "(" + this.x + ", " + this.y + ")";
        };

        this.constructor.prototype.fulfill_target = function() {
            return this.x == TARGET_VOLUME || this.y == TARGET_VOLUME;
        };

        this.constructor.prototype.fill_x = function() {
            return new this.constructor(VOLUME_X, this.y);
        };

        this.constructor.prototype.fill_y = function() {
            return new this.constructor(this.x, VOLUME_Y);
        };

        this.constructor.prototype.empty_x = function() {
            return new this.constructor(0, this.y);
        };

        this.constructor.prototype.empty_y = function() {
            return new this.constructor(this.x, 0);
        };

        this.constructor.prototype.pour_x_y = function() {
            var new_y = Math.min(VOLUME_Y, this.x + this.y);
            var new_x = this.x - (new_y - this.y);
            return new this.constructor(new_x, new_y);
        };

        this.constructor.prototype.pour_y_x = function() {
            var new_x = Math.min(VOLUME_X, this.x + this.y);
            var new_y = this.y - (new_x - this.x);
            return new this.constructor(new_x, new_y);
        };
    }
}

function CupState(steps, parent) {
    this.steps = steps;
    this.parent = parent;
}

var cup_states_container = {
    child_state: function(parent) {
            if (typeof this[parent] == 'undefined') {
                return null;
            }

            return new this[parent].constructor(this[parent].steps + 1, parent);
        }
};
cup_states_container[new Cups(0 ,0)] = new CupState(0, null);
var bfs_queue = [new Cups(0, 0)];
var mutation_methods = ["fill_x", "fill_y", "empty_x", "empty_y", "pour_x_y", "pour_y_x"];

while (bfs_queue.length != 0) {
    var cur_cups = bfs_queue.shift();
    if (cur_cups.fulfill_target()) {
        console.log("Found an answer with " + cup_states_container[cur_cups].steps + " steps!");
        var ans_arr = []
        while (cur_cups) {
            ans_arr.push(cur_cups);
            cur_cups = cup_states_container[cur_cups].parent;
        }
        ans_arr.reverse();
        for (var i = 0; i < ans_arr.length; i++) {
            console.log(ans_arr[i].toString());
        }
        break;
    }
    mutation_methods.forEach(function(sym) {
        var next_cups = cur_cups[sym]();
        if (typeof cup_states_container[next_cups] == 'undefined') {
            cup_states_container[next_cups] = cup_states_container.child_state(cur_cups);
            bfs_queue.push(next_cups);
        }
    });
}

File: cups.rb
VOLUME_X = ARGV[0] ? ARGV[0].to_i : 4
VOLUME_Y = ARGV[1] ? ARGV[1].to_i : 9
TARGET_VOLUME = ARGV[2] ? ARGV[2].to_i : 2

class NilClass
  def clone
    nil
  end
end

class Cups
  attr_reader :x, :y

  def initialize(x = 0, y = 0)
    @x = x
    @y = y
  end

  def ==(another)
    x == another.x and y == another.y
  end

  def eql?(another)
    self == another
  end

  def hash
    [x, y].hash
  end

  def qualify?
    @x == TARGET_VOLUME or @y == TARGET_VOLUME
  end

  def fill_x
    self.class.new(VOLUME_X, y)
  end

  def fill_y
    self.class.new(x, VOLUME_Y)
  end

  def empty_x
    self.class.new(0, y)
  end

  def empty_y
    self.class.new(x, 0)
  end

  def pour_x_y
    new_y = [VOLUME_Y, x+y].min
    new_x = x - (new_y - y)
    self.class.new(new_x, new_y)
  end

  def pour_y_x
    new_x = [VOLUME_X, x+y].min
    new_y = y - (new_x - x)
    self.class.new(new_x, new_y)
  end
end

class CupState
  attr_reader :step, :parent

  def initialize(step, parent)
    @step = step
    @parent = parent.clone
  end
end

cup_states = {}
cup_states[Cups.new(0, 0)] = CupState.new(0, nil)
bfs_queue = [Cups.new(0, 0)]
mutation_methods = [:fill_x, :fill_y, :empty_x, :empty_y, :pour_x_y, :pour_y_x]

cup_states.define_singleton_method(:child) do |cup|
  cup_state = self[cup]
  cup_state.class.new(cup_state.step + 1, cup)
end

until bfs_queue.empty?
  cur_cup = bfs_queue.shift

  if cur_cup.qualify?
    now = cur_cup
    seq = [now]
    until cup_states[now].parent.nil?
      now = cup_states[now].parent
      seq.push now
    end
    seq.reverse!
    puts "Found an answer with #{seq.length - 1} steps!"
    seq.each { |s| puts "(#{s.x}, #{s.y})"}
    break
  end

  mutation_methods.each do |sym|
    new_cup = cur_cup.send(sym)
    next unless cup_states[new_cup].nil?
    cup_states[new_cup] = cup_states.child(cur_cup)
    bfs_queue.push(new_cup)
  end
end
